[HCL] Fix CST-to-Generic translation of heredocs

IagoAbal opened the pull request:
Heredocs are parsed as a list of tokens without any newline or
whitespace. But heredocs may represent code that we want to analyze with
metavariable-regex or metavariable-pattern, so we must parse these
strings correctly. Ideally, we should preserve these newlines and
whitespace already during parsing... but for now we just "restore" them
using token location info.

Closes #4582

test plan:
make test # test included

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

emjin commented on semgrep-core/src/parsing/tree_sitter/Parse_hcl_tree_sitter.ml:
```diff
@@ -85,16 +85,38 @@ let map_numeric_lit (env : env) (x : CST.numeric_lit) : literal =
-      let toks = List.map snd ys in
-      Some (s1 ^ str, PI.combine_infos t1 toks)
+  | ((x_pos, x_str) as x) :: xs ->
+      (* each character is a separate template_literal_chunk, so we need
```
This is an old comment, but just noting that I'm confused by what "each character is a separate template_literal_chunk" means

emjin commented on semgrep-core/src/parsing/tree_sitter/Parse_hcl_tree_sitter.ml:
```diff
@@ -85,16 +85,38 @@ let map_numeric_lit (env : env) (x : CST.numeric_lit) : literal =
+              if x_loc.Loc.start.row > prev_row then base_col else prev_col
+            in
+            let acc_str =
+              (* we need to respect newlines and spaces so that HEREDOCs are
```
This is great, might be better as a function comment so a reader can understand what map_template_string needs to do without looking into the code

emjin reviewed (approved):
Thanks!

IagoAbal committed:
[HCL] Fix CST-to-Generic translation of heredocs

Heredocs are parsed as a list of tokens without any newline or
whitespace. But heredocs may represent code that we want to analyze with
metavariable-regex or metavariable-pattern, so we must parse these
strings correctly. Ideally, we should preserve these newlines and
whitespace already during parsing... but for now we just "restore" them
using token location info.

Closes #4582

test plan:
$ semgrep-core -lang hcl -dump_ast tests/OTHER/rules/terraform_nested_yaml.tf
 #^ see how the definition of content is represented in Generic
$ make test # test included

IagoAbal committed:
Emma's review
